Adding Simulation Evaluation Framework

gwaygenomics opened the pull request:
Related to #103 

Completely updating old method of simulating data - now using WGCNA `simulateDatExpr`

I've also added four evals, which include:

1. Mean ranks of module genes assigned to compressed features (modules have decreasing number of ground truth genes)
2. Mean rank of the list in 1 (except the `noise` module ranks)
3. Euclidean distance reconstruction cost following LSA (A - B + D = C_hat) (`dist(C, decoder(C_hat))`) where A, B, and D are mean latent space encodings of 3 groups of samples. Groups B and D lack gene module 2, Groups A and C have gene module 2. 
4.  LSA isolation of compressed features detecting module 2 in above sample. In effect, testing if subtraction of B from A isolates "essence" of gene module 2.

I have also added a `verbose` argument to the `models` class which will control how training metrics are output.

b2f2bba partially addresses #13 

This is only the framework, results to come in future PR!

gwaygenomics commented:
also tagging @huqiwen0313 for possible thoughts on simulation evals

jaclyn-taroni commented on tybalt/data_models.py:
```diff
@@ -241,3 +287,205 @@ def combine_models(self, include_labels=False, include_raw=False):
+        mean_feature = group_means.loc[group, :]
+        return subtraction + mean_feature
+
+    def reconstruct_group(self, lsa_result, algorithm=False):
```
Should this throw an error (or warning) if `algorithm` is unsupported?

jaclyn-taroni replied:
Same comment as before, looks like @danich1 agrees

jaclyn-taroni commented on tybalt/models.py:
```diff
@@ -105,15 +108,31 @@ def _connect_layers(self):
         self.decoder = Model(decoder_input, _x_decoded_mean)
 
-    def train_vae(self, train_df, test_df):
+    def train_vae(self, train_df, test_df, separate_loss=False):
```
I am not particularly qualified to comment on python conventions, but I would want more information about what `separate_loss` is doing here

jaclyn-taroni commented on tybalt/models.py:
```diff
@@ -254,8 +276,8 @@ class Adage(BaseModel):
     def __init__(self, original_dim, latent_dim, noise=0.05, batch_size=50,
-                 epochs=100, sparsity=0, learning_rate=1.1,
-                 loss='mse'):
+                 epochs=100, sparsity=0, learning_rate=0.0005,
```
Why the change in learning rate (just curious)?

gwaygenomics replied:
I was initially using `Adadelta` as the optimizer to train ADAGE. This worked well after a parameter sweep, but i found that it tended to work poorly with default parameters. `Adam`, which is the optimizer I use for Tybalt, is much more well behaved with different data input under default parameters. For the simulation, I was finding ADAGE to have uniformly poor performance, which was concerning. After some troubleshooting I discovered what I described above and results have improved. I want to compare algorithms, not optimizers here so I chose to keep them consistent - although this is definitely a limitation of comparing non-convex optimization driven algorithms to things like PCA!

jaclyn-taroni commented on tybalt/data_models.py:
```diff
@@ -241,3 +287,205 @@ def combine_models(self, include_labels=False, include_raw=False):
         all_df = pd.concat(all_models, axis=1)
         return all_df
+
+    def get_modules_ranks(self, weight_df, num_components, noise_column=0):
```
Nice documentation on this one

gwaygenomics commented on scripts/util/simulate_expression.R:
```diff
@@ -4,264 +4,108 @@
-                                     seq(1, ncol(cont_other_df)), sep = "_")
-    cont_other_df <- tibble::as_data_frame(cont_other_df)
-  }
+  # Combine Matrix
```
> Ah is it the leaveOut arg?

Precisely! The though being that this will enable an `A-B+D=C` assessment, asking if the arithmetic captures a module that learns the "left out" module. I will add to docs

gwaygenomics commented on scripts/util/simulate_expression.R:
```diff
@@ -4,264 +4,108 @@
+simulateExpression <- function(n, num_sets, num_genes, num_modules,
+                               background_noise, min_cor, max_cor,
+                               sample_set_A, sample_set_B, mod_prop,
+                               leave_out, leave_out_other) {
```
yeah, good point. I will address in next commit

danich1 commented on tybalt/data_models.py:
```diff
@@ -49,19 +51,24 @@ class DataModel():
             self.df = subset_df
 
+        if gene_modules is not None:
+            self.gene_modules = pd.DataFrame(gene_modules).T
```
What is gene_modules? Is it like a one dimensional label vector?

gwaygenomics commented:
Thanks @jaclyn-taroni and @danich1 !

I have updated commits based on your comments. It is still not yet ready for re-review however, I believe I found a bug in the simulated data script that I will need to test when I get back. I will let you know when it's ready again!

gwaygenomics commented:
Alright! I think i have addressed the previous error in fa194fb3baba2cd5ec856b627c927ab97ad57456 (specifically in lines 108-109). Before, I was simulated only 3 sample sets (by row). This would produce signal, but not exactly what I had intended. Concretely, before the sample set was being used to generate the `eigen_samples` matrix. Generating an eigen matrix with 5 gene modules (but only with 3 "sample_sets") resulted in this sampling:

| Gene Module 1 | GM2 | GM3 | GM4 | GM5 |
| :-------------------: | :----: | :-----: | :-----: | :-----: |
| Sample Set A | B | C | A | B |
| C | A | B | C | A |
| B | C | A | B | C |

Instead, with 5 "sample_sets":

| Gene Module 1 | GM2 | GM3 | GM4 | GM5 |
| :-------------------: | :----: | :-----: | :-----: | :-----: |
| Sample Set A | B | C | D | E |
| A | B | C | D | E |
| A | B | C | D | E |

Where, the rows of `eigen_samples` correspond to samples and columns are gene modules.

Additionally, I added a couple figures in 828333a that describes an example of the simulated data.

Ready for re-review! Thanks! (The results of the sweep will be added in a future pull request)

jaclyn-taroni commented:
@gwaygenomics is there a row and column of missing values in [`figures/example_simulated_data.png`](https://github.com/gwaygenomics/tybalt/blob/98733d6c896c0ffda03bc526f881c4979dd3c700/figures/example_simulated_data.png) or is there just something wonky with the graphics that would be remedied by changing the size of the plot?

jaclyn-taroni commented on scripts/util/simulate_expression.R:
```diff
@@ -4,264 +4,206 @@
+  # num_genes - The number of genes to simulate
+  # num_modules - The number of distinct simulated "groups" of genes
+  # background_noise - N(0, sd = background_noise) added to all values
+  # min_cor - The minimum correlation of genes in modules to core module
```
Does core module mean eigengene?

gwaygenomics replied:
yes, good point. That is more accurate

jaclyn-taroni commented on scripts/util/simulate_expression.R:
```diff
@@ -4,264 +4,206 @@
+background_noises <- c(0, 0.1, 0.5, 1, 3)
+genes <- c(500, 1000)
+
+# Other constants
```
It might be a little "cleaner" to set these as defaults for `simulateExpression` but up to you

gwaygenomics replied:
Yeah, it could be cleaner in the sense of less lines, but I think separating out the parameters from the function makes it a bit easier to read/understand.

gwaygenomics commented:
> is there a row and column of missing values in figures/example_simulated_data.png or is there just something wonky with the graphics that would be remedied by changing the size of the plot?

Just wonky 💩 - I will fix before merging

gwaygenomics commented:
> Same comment as before, looks like @danich1 agrees

> I approve. Just make sure you add an exception statement in the reconstruct_group function.

My bad, I must have just missed that one. Fixed in aa754c2
